controllers/kyma_controller.go

```diff
@@ -90,95 +91,95 @@ type KymaReconciler struct {
-		return r.deleteKyma(ctx, kyma)
+		if err := r.deleteRemoteKyma(ctx, kyma); err != nil {
+			return r.requeueWithError(ctx, kyma, err)
+		}
+		if err := r.updateStatus(ctx, kyma, v1beta2.StateDeleting, "waiting for modules to be deleted"); err != nil {
+			return r.requeueWithError(ctx, kyma, fmt.Errorf("could not update kyma status after triggering deletion: %w", err))
+		}
+		return ctrl.Result{}, nil
 	}
 
-	// check finalizer
-	if kyma.CheckLabelsAndFinalizers() {
+	if needsUpdate := kyma.EnsureLabelsAndFinalizers(); needsUpdate {
 		if err := r.Update(ctx, kyma); err != nil {
-			return r.CtrlErr(ctx, kyma, fmt.Errorf("could not update kyma after finalizer check: %w", err))
+			return r.requeueWithError(ctx, kyma, fmt.Errorf("failed to update kyma after finalizer check: %w", err))
 		}
 		return ctrl.Result{}, nil
 	}
 
-	// update the remote kyma with the state of the control plane
 	if kyma.SyncEnabled() {
-		if err := r.syncRemoteKymaSpecAndStatus(ctx, kyma); err != nil {
-			return r.CtrlErr(ctx, kyma, fmt.Errorf("could not synchronize remote kyma: %w", err))
+		if err := r.syncRemoteKyma(ctx, kyma); err != nil {
+			return r.requeueWithError(ctx, kyma, fmt.Errorf("could not synchronize remote kyma: %w", err))
 		}
 	}
 
-	// state handling
-	return r.stateHandling(ctx, kyma)
+	return r.processKymaState(ctx, kyma)
 }
 
-func (r *KymaReconciler) deleteKyma(ctx context.Context, kyma *v1beta2.Kyma) (ctrl.Result, error) {
-	if err := r.TriggerKymaDeletion(ctx, kyma); err != nil {
-		return r.CtrlErr(ctx, kyma, err)
+func (r *KymaReconciler) deleteRemoteKyma(ctx context.Context, kyma *v1beta2.Kyma) error {
+	logger := ctrlLog.FromContext(ctx).V(log.InfoLevel)
+	if kyma.SyncEnabled() {
+		if err := remote.DeleteRemotelySyncedKyma(ctx, kyma, r.RemoteSyncNamespace); client.IgnoreNotFound(err) != nil {
+			logger.Error(err, "Failed to be deleted remotely!")
+			return fmt.Errorf("error occurred while trying to delete remotely synced kyma: %w", err)
+		}
+		logger.Info("Successfully deleted remotely!")
 	}
+	return nil
+}
 
-	// if the status is not yet set to deleting, also update the status of the control-plane
-	// in the next sync cycle
-	if err := status.Helper(r).UpdateStatusForExistingModules(
-		ctx, kyma, v1beta2.StateDeleting, "waiting for modules to be deleted",
-	); err != nil {
-		return r.CtrlErr(ctx, kyma, fmt.Errorf(
-			"could not update kyma status after triggering deletion: %w", err))
-	}
-	return ctrl.Result{}, nil
+func (r *KymaReconciler) requeueWithError(ctx context.Context, kyma *v1beta2.Kyma, err error) (ctrl.Result, error) {
+	return ctrl.Result{Requeue: true}, r.updateStatusWithError(ctx, kyma, err)
 }
 
-func (r *KymaReconciler) CtrlErr(ctx context.Context, kyma *v1beta2.Kyma, err error) (ctrl.Result, error) {
-	return ctrl.Result{Requeue: true},
-		r.UpdateStatusWithEventFromErr(ctx, kyma, v1beta2.StateError, err)
+func (r *KymaReconciler) enqueueWarningEvent(kyma *v1beta2.Kyma, reason EventReason, err error) {
```

Comment:
I think it make sense to design two types of EventReason, to distinguish warning and normal event, and avoid misuse.

Reply:
There are two functions for that now. Reason is just an informative string. What do you mean by can be misused here?